bedrock/src/secure/mod.rs

```diff
@@ -0,0 +1,356 @@
+    ///
+    /// This function can be used in cases where `RootKey` is decoded regularly (for v0 or v1) and then
+    /// re-encoded to full JSON format. For example, backup service parses the key from HEX/JSON
+    /// on enrollment (using regular `decode`) and then re-encodes it to JSON format for storage.
+    /// When recovery is happening, we know the key is in JSON format, so we can use this function
+    /// to decode it at the recovery time.
+    ///
+    /// # Errors
+    /// Will return an error if the key is not a validly encoded `RootKey`.
+    #[uniffi::constructor]
+    pub fn decode_from_json_enforced(encoded_key: &str) -> Result<Self, SecureError> {
+        serde_json::from_str::<Self>(encoded_key)
+            .map_err(|_| SecureError::KeyParseError)
+    }
+
+    /// Encodes the key as JSON.
+    ///
+    /// # Errors
+    /// Will return an error if the key cannot be serialized to JSON (highly unexpected).
+    pub fn encode(&self) -> Result<String, SecureError> {
+        serde_json::to_string(self).map_err(|_| SecureError::Generic {
+            message: "unable to encode RootKey as JSON".to_string(),
+        })
+    }
+
+    /// Returns true if the key is in the legacy format and don't used indexed derivation.
+    #[must_use]
+    pub fn is_v0(&self) -> bool {
+        matches!(self.key, VersionedKey::V0(_))
+    }
+
+    /// Returns the Ethereum key as hex (without leading 0x).
+    ///
+    /// # Errors
+    /// Will return an error if the key cannot be derived (highly unexpected).
+    pub fn ethereum_key(&self) -> Result<String, SecureError> {
+        // In the old version, the key was used directly.
+        Ok(match &self.key {
+            VersionedKey::V0(str) => hex::encode(derive_key_v0(str.clone())?),
+            VersionedKey::V1(key) => hex::encode(derive_subkey(key, ETHEREUM_KEY_ID)?),
+        })
+    }
+
+    /// Returns the Ethereum key at a given index. The index is used to enable key rotation for Ethereum keys.
+    ///
+    /// The key is returned as hex (without leading 0x).
+    ///
+    /// # Errors
+    /// Will return an error if the key cannot be derived (highly unexpected).
+    pub fn ethereum_key_with_index(&self, index: u64) -> Result<String, SecureError> {
+        Ok(match &self.key {
+            VersionedKey::V0(str) => {
+                if index == 0 {
+                    hex::encode(derive_key_v0(str.clone())?)
+                } else {
+                    // In V0, the base key is used directly for different purposes,
+                    // whereas in V1, the main key serves as a seed for deriving subkeys specific to each purpose.
+                    // To enable predictable generation of a new Ethereum key from an index,
+                    // we must derive a seed key that is distinct from the original one, similar to the "marble" approach.
+                    let mut hasher = Sha256::new();
+                    hasher.update(ROTATED_ETHEREUM_KEY_SALT);
```

Comment:
How does this logic work - won't every user going to get the same derived v0 keys?

Shouldn't we throw an error if we try to derive subkey of v0?